converter/internal/otelcolconvert/otelcolconvert.go

```diff
@@ -0,0 +1,255 @@
+		return nil, diags
+	}
+
+	cfg, err := readOpentelemetryConfig(in)
+	if err != nil {
+		diags.Add(diag.SeverityLevelCritical, err.Error())
+		return nil, diags
+	}
+	if err := cfg.Validate(); err != nil {
+		diags.Add(diag.SeverityLevelError, fmt.Sprintf("failed to validate config: %s", err))
+		return nil, diags
+	}
+
+	f := builder.NewFile()
+
+	diags.AddAll(appendConfig(f, cfg))
+	diags.AddAll(common.ValidateNodes(f))
+
+	var buf bytes.Buffer
+	if _, err := f.WriteTo(&buf); err != nil {
+		diags.Add(diag.SeverityLevelCritical, fmt.Sprintf("failed to render Flow config: %s", err.Error()))
+		return nil, diags
+	}
+
+	if len(buf.Bytes()) == 0 {
+		return nil, diags
+	}
+
+	prettyByte, newDiags := common.PrettyPrint(buf.Bytes())
+	diags.AddAll(newDiags)
+	return prettyByte, diags
+}
+
+func readOpentelemetryConfig(in []byte) (*otelcol.Config, error) {
+	provider := yamlprovider.New()
+
+	configProvider, err := otelcol.NewConfigProvider(otelcol.ConfigProviderSettings{
+		ResolverSettings: confmap.ResolverSettings{
+			URIs: []string{"yaml:" + string(in)},
+			Providers: map[string]confmap.Provider{
+				provider.Scheme(): provider,
+			},
+			Converters: []confmap.Converter{expandconverter.New()},
+		},
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create otelcol config provider: %w", err)
+	}
+
+	cfg, err := configProvider.Get(context.Background(), getFactories())
+	if err != nil {
+		// TODO(rfratto): users may pass unknown components in YAML here. Can we
+		// improve the errors? Can we ignore the errors?
+		return nil, fmt.Errorf("failed to get otelcol config: %w", err)
+	}
+
+	return cfg, nil
+}
+
+func getFactories() otelcol.Factories {
+	facts := otelcol.Factories{
+		Receivers:  make(map[component.Type]receiver.Factory),
+		Processors: make(map[component.Type]processor.Factory),
+		Exporters:  make(map[component.Type]exporter.Factory),
+		Extensions: make(map[component.Type]extension.Factory),
+		Connectors: make(map[component.Type]connector.Factory),
+	}
+
+	for _, converter := range converters {
+		fact := converter.Factory()
+
+		switch fact := fact.(type) {
+		case receiver.Factory:
+			facts.Receivers[fact.Type()] = fact
+		case processor.Factory:
+			facts.Processors[fact.Type()] = fact
+		case exporter.Factory:
+			facts.Exporters[fact.Type()] = fact
+		case extension.Factory:
+			facts.Extensions[fact.Type()] = fact
+		case connector.Factory:
+			facts.Connectors[fact.Type()] = fact
+
+		default:
+			panic(fmt.Sprintf("unknown component factory type %T", fact))
+		}
+	}
+
+	return facts
+}
+
+// appendConfig converts the provided OpenTelemetry config into an equivalent
+// Flow config and appends the result to the provided file.
+func appendConfig(file *builder.File, cfg *otelcol.Config) diag.Diagnostics {
+	var diags diag.Diagnostics
+
+	groups, err := createPipelineGroups(cfg.Service.Pipelines)
+	if err != nil {
+		diags.Add(diag.SeverityLevelCritical, fmt.Sprintf("failed to interpret config: %s", err))
```

Comment:
I'm marking this comment on one of them but it would be nice to see some testing where various diags get hit. It's not 100% obvious what will happen for each of them in terms of whether we get a partially correct .river or not

Reply:
I added some tests for critical errors that can be hit; not all of them are possible to hit and are written defensively. 

I put these in a separate folder so they don't get in the way of the component conversion tests.